iree/compiler/Codegen/SPIRV/KernelConfig.h

```diff
@@ -15,11 +15,46 @@
 #ifndef IREE_COMPILER_CODEGEN_SPIRV_KERNELCONFIG_H_
 #define IREE_COMPILER_CODEGEN_SPIRV_KERNELCONFIG_H_
 
+#include <array>
+
+#include "iree/compiler/Dialect/HAL/IR/LoweringConfig.h"
+#include "mlir/Dialect/SPIRV/IR/TargetAndABI.h"
 #include "mlir/IR/BuiltinOps.h"
 
 namespace mlir {
 namespace iree_compiler {
 
+namespace detail {
+
+struct SPIRVCodeGenConfig {
```

Comment:
This is duplication of information that is handled through attribute. To me this is congnitive overhead. Everytime you need to add something new to this you need to change it in two places instead of just have a single source of truth. I am not really sure what this is meant for.

Reply:
This isn't going back to the old C++ launch config days. Back then we perform duplicate query in each pass and just use it directly in C++. Here it's just for better layering. The whole pipeline is still chained together with `lowering.config` attributes; so that hasn't been changed. The main goal here is to separate the concerns. For per-target configuration, it should only concern about the configuration themselves. Not all the attribute setting etc. That's better to be handled in one single place to avoid duplication and hiding all the magic in deeply nested function calls. Think of policy vs mechanisms.

I think we should improve how these configurations are integrated in IREE in general. I prefer to define some attribute to make them more structured. But for now C++ is fine, as there are other pressing issues to address. Also note that as of today, we are still encoding these in C++ anyway. It's just mixed together with attribute setting code. :)

Reply:
I still dont get this reasoning. There is a little wonkiness w.r.t which attributes go where, but I am not understanding why it is better to have a C++ struct that carries this information (that needs to be passed back through the call stack) to the place where it is set. 

The current utility method is not "magic". Once you decide the configuration for the root op, you know everything that you need to know about the lowering config, the number of workgroups, the workgroup size, etc. So setting at the point where you are deciding is better than making sure that the decided configuration is then propagated back all the way to some place in the call stack to be set. 

Also, this is again now making the SPIR-V code-generation divergent w.r.t CUDA and LLVM CPU backends getting us back to a point where adding something to CPU and CUDA backends is one way and then SPIR-V backend is another way. For example, there are a list of tile sizes below which includes `convFilterTileSizes` which is not needed for any other operation. This should not exist here. THe way to handle convolution is to have a separate pipeline for lowering convolutions, and tile sizes set w.r.t to that pipeline. So when you are deciding the configuration for convolutions, you also set the tile sizes that make sense for that pipeline. It does not need to be parried back through the call stack.

Reply:
I don't think this makes SPIR-V side different from LLVM CPU/GPU as far as how the general mechanism and "adding something" is concerned. It's still using the `lowering.config` attribute to convey the whole information (so not replacing that with C++ constructs). Adding configurations is still via C++ on all the backends. There is no much difference here. What differs is implementation details.

The C++ structure in this patch is actually entirely scoped to KernelConfig.cpp itself; it's _internal implementation details_. Maybe it's easier to think this way: we can treat the `setOpConfigAndEntryPointFnTranslation` call and the workgroup count region definition as just C++ functions. What we do before this change is inlining those functions in every caller. Now this change moves them out for better layering and avoid duplication. Again, it's still entirely internal to KernelConfig.cpp (and hence the `detail` namespace); it's just that I need to split it into multiple files so the definition is placed in a header.

To expand more on why this. One of the big differences between SPIR-V and ARM CPU/CUDA is that SPIR-V is for many vendors and architectures. So by definition it needs to handle all those complexity of different vendors and architectures, compared to the simplicity (and single vendor/architecture) for ARM CPU/CUDA. The solution should fit the problem space. So I think it makes sense on the SPIR-V side the implementation differs from ARM CPU/CUDA, because it have _different needs_. It's desirable to have a consistent general mechanism for all backends; but I think it's hard to require all backends to maintain the same exact implementation details.

Now with some concrete numbers. Due to single vendor/architecture, in ARM CPU / CUDA side we have  only 4 / 3 calls of `setOpConfigAndEntryPointFnTranslation`. That's entirely fine to inline! If we have similar situation on SPIR-V side I'd happy do the exact same. But on SPIR-V side, next step I will add basic Adreno default configuration and update/fix the NV default configuration. With Mali, there are 3 vendors (this is without even considering the default for no vendor match). Even for the 4 ops we support (matmul, batch matmul, conv_2d, dw_conv_2d), that's 3 x 4 = 12 callers have it inlined. Also for convolution, right now we need the special treatment of static workgroup count setting. So that's 3 x 2 = 6 places have the same logic duplicated. I feel hard to understand why 12 + 6 = 18 duplications is better than 1 + 1 = 2 in one single place. And the duplication can only getting worse as we support more architectures and ops.

By moving it out to one place, due to the separate of concerns, I'd actually argue it also makes updating either side easier. Say if you are changing the signature, there is just one place to replace, instead more than 10. Adding new configurations (either Thomas or me) will not need to copy over all the attribute and workgroup region setting code.

The weirdness of `convFilterTileSizes` is a good point. Actually that should be directly placed in the pipeline, as it's not tunable  knobs that adjust with workload size; it's actually always the same configuration to satisfy constraints to make following patterns kick in. So it's fundamentally different than other levels of tiling. I can make the change later (or in this patch directly).

This went quite long. :) Let me recap key points:

* It's not using C++ to replace the existing attribute for pipeline lowering control. The overall mechanism stays the same.
* It's only changing the internal implementation of KernelConfig.cpp to make it more structured and separated.
* We need this because of the problem SPIR-V faces: multiple vendors/architectures.
* Because of the separate concerns, it's easier to read and update.

We can sync via GVC if this is still not clear.  :)

Reply:
I still dont see the reasoning here of why there insistence on having this c++ structure created an parlayed back to the caller. I am fine with splitting things up into different files. If the existing utility methods dont work for SPIR-V, would be good to know why, and would also be fine to add new utility methods if that helps. Nothing here is convincing me about having yet another c++ struct to carry the configuration information.